app/src/main/java/org/kiwix/kiwixmobile/webserver/WebServerHelper.java

```diff
@@ -0,0 +1,117 @@
+package org.kiwix.kiwixmobile.webserver;
+
+import android.content.Context;
+import android.util.Log;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import org.kiwix.kiwixlib.JNIKiwixException;
+import org.kiwix.kiwixlib.JNIKiwixLibrary;
+import org.kiwix.kiwixlib.JNIKiwixServer;
+
+/**
+ * WebServerHelper class is used to set up the suitable environment i.e. getting the
+ * ip address and port no. before starting the WebServer
+ * Created by Adeel Zafar on 18/07/2019.
+ */
+
+public class WebServerHelper {
+  Context context;
+  public static boolean isServerStarted;
+  static int port;
+  JNIKiwixLibrary kiwixLibrary = new JNIKiwixLibrary();
+  JNIKiwixServer kiwixServer = new JNIKiwixServer(kiwixLibrary);
+  private static final String TAG = "WebServerHelper";
+
+  public WebServerHelper(Context context) {
+    this.context = context;
+  }
+
+  public boolean startServerHelper(ServerStateListener stateListener,
+      ArrayList<String> selectedBooksPath) {
+
+    // 1. Get port from settings screen
+    // 2. Ask user to change port in settings if port is in use.
+    // OR
+    // Always use 8080 and when its not available then iterate this number.
+    String ip = getIpAddress();
+    ip = ip.replaceAll("\n", "");
+    if (ip.length() == 0) {
+      stateListener.serverFailed();
+    } else if (!isServerStarted && startAndroidWebServer(selectedBooksPath)) {
+      stateListener.serverStarted("http://" + ip + ":" + port);
+    }
+    return isServerStarted;
+  }
+
+  public boolean stopAndroidWebServer(ServerStateListener stateListener) {
+    if (isServerStarted) {
+      kiwixServer.stop();
+      isServerStarted = false;
+      stateListener.serverStopped();
```

Comment:
@macgills This stateListener object is thrown as null.

Reply:
Okay, do not pass your listener around, this listener belongs to HotspotService, you register it there so keep it there.
You will have to check that the listener is not null before you invoke methods on it.
You bind/unbind in start/stop, also you register/unregister the listener at these points. That means that if you press the home button (leave ZimHostActivity in any way) the listener gets set to null. Now if `stopHotspotAndDismissNotification();` gets invoked you have a NPE from `ACTION_STOP` `ACTION_TURN_OFF_AFTER_O`, `ACTION_STOP_SERVER`

Reply:
Yep, you're right!

Reply:
I've added null checks in these classes. `serverstateListener` can't be thrown null now.